### PR TITLE
Introduce revocation endpoint

### PR DESCRIPTION
**Pull request by ctriant**

### CODE REVIEW

**Comment by rohe** [comment]
I think you should rename the endpoint class to TokenRevokation.

I also want to see lots of tests :-)

**Comment by ctriant** [comment]
> I think you should rename the endpoint class to TokenRevokation.
> 
Are you sure about that? I think the word is revocation with 'c'. It's used that way in the spec too.

> I also want to see lots of tests :-)

I forgot to push the tests file. O:)  Anyway, I'm working on some changes locally and I'll rebase and push once ready for review.

**Comment by rohe** [comment]
TokenRevocation instead of just Revocation.

File: doc/server/contents/conf.rst [hunk 1]
```diff
@@ -837,3 +850,102 @@ For example::
...
+            "kwargs": {
+              "scopes": ["openid"]
+            }
+          }
+        }
+    }
+}
+
+For the per-client configuration a similar configuration scheme should be present in the client's
+metadata under the `token_exchangerevocation` key.
```
Comment (c00kiemon5ter) [inline doc/server/contents/conf.rst]:
```suggestion
metadata under the `token_revocation` key.
```

File: doc/server/contents/conf.rst [hunk 2]
```diff
@@ -837,3 +850,102 @@ For example::
...
         return request
 
+==============
+Token revocation
+==============
+        
+In order to enable the token revocation endpoint a dictionary with key `token_revocation` should be placed 
+under the `endpoint` key of the configuration.
+
+If present, the token revocation configuration should contain a `policy` dictionary
```
Comment (c00kiemon5ter) [inline doc/server/contents/conf.rst:971]: Is the `policy` key needed? or is a policy applied by default when the key is missing?

```suggestion
If present, the token revocation configuration may contain a `policy` dictionary
```

File: doc/server/contents/conf.rst [hunk 3]
```diff
@@ -837,3 +850,102 @@ For example::
...
+            "kwargs": {
+              "scopes": ["openid"]
+            }
+          }
+        }
+      }
+    }
+
+The policy callable accepts a specific argument list and handles the revocation appropriately and returns 
+a oauth2.TokenRevocationResponse or raises an exception.
```
Comment (c00kiemon5ter) [inline doc/server/contents/conf.rst]:
```suggestion
an :py:class:`idpyoidc.message.oauth2.TokenRevocationResponse` or raises an exception.
```

**Comment by rohe** [comment]
I will add this PR to the fedservice branch. Should allow us to kill this PR.

**Comment by rohe** [comment]
Merged it into the foodservice branch and the tests run OK.
So, the basics are working. Will check for non-obvious things now :-)

**Comment by rohe** [comment]
The token revocation endpoint URL is not supposed to be published as part of the server metadata.
How can this be regarded as a security feature ?? Security by obscurity ??

**Comment by c00kiemon5ter** [comment]
I am not sure what you mean @rohe.

The token revocation endpoint is part of the OAuth2 server metadata (RFC8414).

Reference: https://www.rfc-editor.org/rfc/rfc8414.html#section-2

```
revocation_endpoint
      OPTIONAL.  URL of the authorization server's OAuth 2.0 revocation
      endpoint [[RFC7009](https://www.rfc-editor.org/rfc/rfc7009)].

   revocation_endpoint_auth_methods_supported
      OPTIONAL.  JSON array containing a list of client authentication
      methods supported by this revocation endpoint.  The valid client
      authentication method values are those registered in the IANA
      "OAuth Token Endpoint Authentication Methods" registry
      [[IANA.OAuth.Parameters](https://www.rfc-editor.org/rfc/rfc8414.html#ref-IANA.OAuth.Parameters)].  If omitted, the default is
      "client_secret_basic" -- the HTTP Basic Authentication Scheme
      specified in [Section 2.3.1](https://www.rfc-editor.org/rfc/rfc8414.html#section-2.3.1) of OAuth 2.0 [[RFC6749](https://www.rfc-editor.org/rfc/rfc6749)].

   revocation_endpoint_auth_signing_alg_values_supported
      OPTIONAL.  JSON array containing a list of the JWS signing
      algorithms ("alg" values) supported by the revocation endpoint for
      the signature on the [JWT](https://www.rfc-editor.org/rfc/rfc8414.html#ref-JWT) [JWT] used to authenticate the client at
      the revocation endpoint for the "private_key_jwt" and
      "client_secret_jwt" authentication methods.  This metadata entry
      MUST be present if either of these authentication methods are
      specified in the "revocation_endpoint_auth_methods_supported"
      entry.  No default algorithms are implied if this entry is
      omitted.  The value "none" MUST NOT be used.
```

**Comment by rohe** [comment]
It's me (grumpy old man) having a problem with the sentence in RFC7009 that says: "Since the revocation endpoint is handling security credentials,  clients need to obtain its location from a trustworthy source only."

I'm not sure I would regard normal discovery as completely trustworthy. :-)

**Comment by rohe** [comment]
Moved to the fedservice branch so removing it here.

**Comment by peppelinux** [comment]
I agree with you on the "value" and the meaning of truswothness!

**Pull request closed without merging**